app/main.c

```diff
@@ -369,13 +381,53 @@ static void setup_vm_map(uintptr_t win_start, uintptr_t win_end)
             seg_end = win_end;
         }
         if (seg_start < seg_end && seg_start < win_end && seg_end > win_start) {
-            num_mapped_pages += seg_end - seg_start;
-            vm_map[vm_map_size].pm_base_addr = seg_start;
-            vm_map[vm_map_size].start        = first_word_mapping(seg_start);
-            vm_map[vm_map_size].end          = last_word_mapping(seg_end - 1, sizeof(testword_t));
-            vm_map_size++;
+            // We need to test part of that physical memory segment.
+            if (!enable_numa) {
+non_numa_vm_map_entry:
+                num_mapped_pages += seg_end - seg_start;
+                vm_map[vm_map_size].pm_base_addr = seg_start;
+                vm_map[vm_map_size].start        = first_word_mapping(seg_start);
+                vm_map[vm_map_size].end          = last_word_mapping(seg_end - 1, sizeof(testword_t));
+                vm_map_size++;
+            }
+            else {
+                // Now also pay attention to proximity domains, which are based on physical addresses.
+                uintptr_t orig_start = seg_start << PAGE_SHIFT;
+                uintptr_t orig_end = seg_end << PAGE_SHIFT;
+                uint32_t proximity_domain_idx;
+                uintptr_t new_start;
+                uintptr_t new_end;
+
+                while (1) {
+                    if (smp_narrow_to_proximity_domain(orig_start, orig_end, &proximity_domain_idx, &new_start, &new_end)) {
```

Comment:
How often does a segment straddle proximity domains (given that the segment size is limited by the 1GB window)? If that is a rare case, this could be made much simpler by just allowing all CPUs whose proximity regions intersect the segment to work on that segment.

Reply:
All I know is that I tried to play safe, and that I'm pretty sure I debugged through that code path on QEMU :)
The prime contributor to 1 GB segments straddling across proximity domains is probably the memory hole.